### DISCUSSION TITLE
Answer streaming

### DISCUSSION THREAD

**Discussion by gyzerok** [open-webui/open-webui#2142] [Q&A] [answered]
Hello and thank you for such a great project!

I've looked through the settings and does not see any ability to enable streaming of the model response. I'd like to see tokens appearing as the model generates them. Is it not a feature available yet?

**Comment by justinh-rahb** [comment, marked as answer]
This has always been a feature and is the default behaviour. Various proxy and/or tunnelling setups have been found to interfere with this however.

**Reply by gyzerok** [reply]
Oh, thank you! I am using Caddy, so it's probably about it. Will research it. If you happen to know  what to do, I will appreciate yor advice.

Is it done with websockets or sse on the open-webui side?

**Reply by justinh-rahb** [reply]
SSE

**Reply by clang88** [reply]
If you are using a proxy like nginx, I found that enabling http/2 support also enables the answer streaming.

**Reply by christopherquenneville** [reply]
> If you are using a proxy like nginx, I found that enabling http/2 support also enables the answer streaming.

You just saved me hours of troubleshoot, thank you very much

**Comment by krausality** [comment]
Alright, so i needed an https proxy for enabling the voice input feature on my smartphone, which connects to a local server Open WebUI running on windows pc.
Ollama via the windows installer was already setup previous to this.

This was my initial nginx setup:

`mkdir nginx_proxy_server`
`cd nginx_proxy_server`

1. Create the `Dockerfile`:

```Dockerfile
FROM nginx:alpine

# Install OpenSSL
RUN apk add --no-cache openssl

# Generate self-signed certificate for both localhost and IP
RUN openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/nginx/cert.key -out /etc/nginx/cert.crt -subj "/CN=localhost" -addext "subjectAltName = IP:192.168.178.102, DNS:localhost"

# Copy nginx configuration
COPY nginx.conf /etc/nginx/nginx.conf

# Expose ports
EXPOSE 80 443

# Start Nginx
CMD ["nginx", "-g", "daemon off;"]
```

2. Create the `nginx.conf` file:

```nginx
events {
    worker_connections 1024;
}

http {
    server {
        listen 80;
        server_name localhost 192.168.178.102;
        return 301 https://$host$request_uri;
    }

    server {
        listen 443 ssl;
        server_name localhost 192.168.178.102;

        ssl_certificate /etc/nginx/cert.crt;
        ssl_certificate_key /etc/nginx/cert.key;

        location / {
            proxy_pass http://host.docker.internal:3000;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
        }
    }
}
```

3. Create `docker-compose.yml` file:

```yaml
version: '3.8'

services:
  nginx-ssl-proxy:
    build: .
    ports:
      - "80:80"
      - "443:443"
    extra_hosts:
      - "host.docker.internal:host-gateway"
    restart: always  # This is the equivalent of --restart always

```

4. Build the container:

```powershell
docker-compose up -d --build
```

5. Ensure your Open WebUI container is running:

```powershell
docker run -d -p 3000:8080 --name open-webui --restart always ghcr.io/open-webui/open-webui:main
```

This solved my inital microphone problem BUT **now i had issues with the Answer streaming**.

So all i did was the following:

1. Modify the nginx.conf located in the nginx_proxy_server to the following version:

```
events {
    worker_connections 1024;
}
http {
    map $http_accept $sse_content_type {
        "~*text/event-stream" text/event-stream;
        default application/json;
    }
    server {
        listen 80;
        server_name localhost 192.168.178.102;
        return 301 https://$host$request_uri;
    }
    server {
        listen 443 ssl;
        ssl_protocols TLSv1.2 TLSv1.3;
        ssl_prefer_server_ciphers off;
        server_name localhost 192.168.178.102;
        ssl_certificate /etc/nginx/cert.crt;
        ssl_certificate_key /etc/nginx/cert.key;

        # Set the maximum allowed size of the client request body
        client_max_body_size 2000M;  # Adjust this value based on your needs

        # Configure SSE only for specific endpoints
        location ~* ^/api/chat/completed|/api/v1/chats|/api/task/title/completions|/ollama/api/chat {
            proxy_pass http://host.docker.internal:3000;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            # Enable SSE settings for these endpoints
            proxy_set_header Connection '';
            proxy_http_version 1.1;
            chunked_transfer_encoding off;
            proxy_buffering off;
            proxy_cache off;
            proxy_set_header Content-Type $sse_content_type;
            # High timeouts for streaming
            proxy_read_timeout 24h;
            proxy_send_timeout 24h;
            keepalive_timeout 24h;
        }
        # Default settings for all other endpoints
        location / {
            proxy_pass http://host.docker.internal:3000;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_read_timeout 60s;
            proxy_send_timeout 60s;
        }
    }
}

```

2. Rebuild and restart the container:

```
docker-compose down
docker-compose up -d --build
```

That worked for me! Now answer streaming is back again :)

*EDIT* Fixed some bugs and edited the files accordingly. Further fixes, issues etc will be tracked here: https://github.com/krausality/nginx-https-open-webui-proxy

**Reply by fcarq** [reply]
> 
>     1. Modify the nginx.conf located in the nginx_proxy_server to the following version:
> 
> 
> ```
> events {
>     worker_connections 1024;
> }
> 
> http {
> 
>     map $http_accept $sse_content_type {
>         "~*text/event-stream" text/event-stream;
>         default application/json;
>     }
>     server {
>         listen 80;
>         server_name localhost 192.168.178.102;
>         return 301 https://$host$request_uri;
>     }
> 
>     server {
>         listen 443 ssl;
>         ssl_protocols TLSv1.2 TLSv1.3;
>         ssl_prefer_server_ciphers off;
>         server_name localhost 192.168.178.102;
> 
>         ssl_certificate /etc/nginx/cert.crt;
>         ssl_certificate_key /etc/nginx/cert.key;
> 
>         location / {
>             proxy_pass http://host.docker.internal:3000;
>             proxy_set_header Host $host;
>             proxy_set_header X-Real-IP $remote_addr;
> 
>             # SSE support
>             proxy_set_header Connection '';
>             proxy_http_version 1.1;
>             chunked_transfer_encoding off;
>             proxy_buffering off;
>             proxy_cache off;
>             proxy_set_header Content-Type $sse_content_type;
> 
>             # Increase timeouts for long-running requests
>             proxy_read_timeout 24h;
>             proxy_send_timeout 24h;
>             keepalive_timeout 24h;
>         }
>     }
> }
> ```
> 
>     2. Rebuild and restart the container:
> 
> 
> ```
> docker-compose down
> docker-compose up -d --build
> ```
> 
> That worked for me! Now answer streaming is back again :)

I was having a strage issue where when using llama3 everything was working, but when using llama3:70b openwebui wans't able to output any response and this was because of my NPM proxy config. I was using npm to access ollama via a local url, not IP, and this was creating an issue with slower responses! With your nginx config it worked! Thank you!

**Reply by krausality** [reply]
I ran into some bugs, fixed them all here: https://github.com/krausality/nginx-https-open-webui-proxy
For including automatic restart swap your docker-compose.yml and for bugs fixing swap the nginx.conf with the ones provided in my repo.
Will also update my initial answer above. Cheers

**Comment by tristan-k** [comment]
@fcarq Can you explain further what to change? I'm also running Nginx-Proxy-Manager and after accessing my Ollama instance through Nginx-Proxy-Manager with Bearer Token I'm no longer able to properly stream chat responses. At the same time it's [not possible](https://github.com/open-webui/open-webui/discussions/4065#discussioncomment-10940517) to disable `Stream Chat Response` globally.

**Comment by vedhavyas** [comment]
For anyone using `caddy-docker-proxy`, the following lables worked for me. 

```
      caddy: example.com
      # remove if not using local certs
      caddy.tls: internal
      # Named matcher for SSE by Accept header
      caddy.@sse.header: "Accept text/event-stream"
      # Handle SSE first
      caddy.handle_0: "@sse"
      caddy.handle_0.header.Cache-Control: "no-cache"
      caddy.handle_0.reverse_proxy: "{{upstreams 8080}}"
      caddy.handle_0.reverse_proxy.flush_interval: "-1"
      # Fallback: normal traffic (compression allowed)
      caddy.handle_1: ""
      caddy.handle_1.encode: "zstd gzip"
      caddy.handle_1.reverse_proxy: "{{upstreams 8080}}"
```      

Let me know if there is anything missing in the labels I should include
